storage/index.js

```diff
@@ -1,20 +1,56 @@
+var createStorageBackend = require('./backend')
+
 /**
  * Storage wrapper object for relaying operations to backend.
  */
-module.exports = function (backend) {
+module.exports = function (STATE, URL) {
+  var backend = undefined
+  var backendCache = {}
+
+  // TODO add some clean up functionality in update for cached backends
+  // that no longer should stay alive to clean up resources etc
+
+  function createNewBackend (state, url) {
+    var debug = state.params.debug
+    backend = createStorageBackend(state, url)
+    if (backend) {
+      if (debug) console.log('storage: created new backend for url', url)
+      backendCache[url] = backend
+    } else {
+      console.warn('storage: missing protocol handler for url', url)
+    }
+  }
+
+  function updateBackend (state, url) {
+    var debug = state.params.debug
+    if (backend) {
+      var cached = backendCache[url]
+      if (cached && cached === backend) {
+        if (debug) console.log('storage: not updating backend')
+      } else if (cached) {
+        if (debug) console.log('storage: updating to cached backend for url', url)
+        backend = cached
+      } else if (!cached) {
+        createNewBackend(state, url)
+      }
+    } else {
+      createNewBackend(state, url)
+    }
+  }
+
   var storageFn = function (name) {
     return {
       write: function (offset, buf, cb) {
-        cb(new Error('write not implemented'))
+        throw new Error('write not implemented')
```

Comment:
These are not user errors. They should not happen.

Reply:
Actually, `storageFn` should be able to return a memoized object instead of a new object each time. Maybe we don't call this function that often anyway so it doesn't matter.